### PR TITLE
fix `/mayastor` page table columns and updated id and startedTime

### PR DESCRIPTION
**Pull request by bhaskarhc**
### fixed startedOn, status, jiva,jenkins and ID on mayastor ci pipelines table .
Screenshot 2021-08-16 at 12 41 48 PM

### /mayastor/{jenkins-test-id} navigation issue fixed and updated .
Screenshot 2021-08-16 at 12 42 11 PM

### added /redirectTo/{URL} route to open permission to `open` or`goBack`

Screenshot 2021-08-16 at 12 41 56 PM

### CODE REVIEW

File: src/app/components/doughnut-graph/doughnut-graph.component.html [hunk 1]
```diff
@@ -1,31 +1,48 @@
...
-          stroke-dashoffset="22.5">
-        </circle>
-        <circle class="donut-segment" cx="21" cy="21" r="15.91549430918954" fill="transparent" stroke="#009400"
-          stroke-width="7" [attr.stroke-dasharray]="getJobPercentForPipelineDoughnut(pipeline,'success')"
-          stroke-dashoffset="22.5">
-        </circle>
-
-      </svg>
-    </ng-template>
+<div *ngIf="engine;else doughnutForOpenebsEngine">
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.html:1]: Is this expression correct ? *ngif...else ?

File: src/app/components/doughnut-graph/doughnut-graph.component.html [hunk 2]
```diff
@@ -1,31 +1,48 @@
...
+  <span ngbTooltip="{{ genMayastorTooltip(pipeline)}}" placement="right">
+    <svg height="30px" viewBox="0 0 42 42" class="donut">
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.html:3]: Why inline style attrs ? Can we do it through external stylesheet ?

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 3]
```diff
@@ -23,44 +24,25 @@ export class DoughnutGraphComponent implements OnInit {
       } else if (data.status == "none") {
         return ""
       } else {
-        var passedJobs = this.jobStatusCount(data.jobs,'success')
-        var failedJobs = this.jobStatusCount(data.jobs,'failed')
-        var canceledJobs = this.jobStatusCount(data.jobs,'canceled')
+        const passedJobs = this.jobStatusCount(data.jobs, 'success')
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:27]: We have enums for these strings . Lets pass enums.

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 4]
```diff
@@ -23,44 +24,25 @@ export class DoughnutGraphComponent implements OnInit {
...
+        const failedJobs = this.jobStatusCount(data.jobs, 'failed')
+        const canceledJobs = this.jobStatusCount(data.jobs, 'canceled')
         return `Success: ${passedJobs},\nFailed: ${failedJobs},\nCanceled: ${canceledJobs}`;
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:30]: No need of \n. String literals can span multiple lines. We can utilize that.

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 5]
```diff
@@ -131,5 +113,30 @@ export class DoughnutGraphComponent implements OnInit {
     }
   }
 
+  getStatusMayastor(tests) {
+    try {
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:117]: Do we need try...catch here ?

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 6]
```diff
@@ -131,5 +113,30 @@ export class DoughnutGraphComponent implements OnInit {
...
+      const passedCount = tests.filter((res) => res.status === 'PASSED').length;
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:118]: Let's bring the string from enum. We have that.

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 7]
```diff
@@ -131,5 +113,30 @@ export class DoughnutGraphComponent implements OnInit {
...
+      const totalTestsCount = tests.length;
+      const percentage = (passedCount / totalTestsCount) * 100;
+      if (passedCount === 0) {
+        return `0 0`;
+      }
+      return `${percentage} ${100 - percentage}`;
+    } catch (error) {
+      console.log('\t Error', error);
+      return `100 0`;
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:127]: Why on error we return 100 ?

File: src/app/components/doughnut-graph/doughnut-graph.component.ts [hunk 8]
```diff
@@ -131,5 +113,30 @@ export class DoughnutGraphComponent implements OnInit {
...
+    }
+  }
+  genMayastorTooltip(tests) {
+    try {
```
Comment (satyapriyamishra222) [inline src/app/components/doughnut-graph/doughnut-graph.component.ts:131]: try...catch & raw strings.

File: src/app/components/external-link-access/external-link-access.component.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<div class="container-fluid ">
+    <div class="w-50 mx-auto border shadow rounded bg-white p-3">
+        <div class="warning">
+            <h5 class="message text-danger"> you're being redirected to internal Mayadata system</h5>
```
Comment (satyapriyamishra222) [inline src/app/components/external-link-access/external-link-access.component.html:4]: Let's do translation please.

File: src/app/components/external-link-access/external-link-access.component.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+            <p class="url text-secondary">{{link}}</p>
+        </div>
+        <div class="action-button d-flex justify-content-around">
+            <div class="open">
+                <a class="btn btn-primary" href="{{link}}">
+                    open
```
Comment (satyapriyamishra222) [inline src/app/components/external-link-access/external-link-access.component.html:10]: translation

File: src/app/components/external-link-access/external-link-access.component.html [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+                </a>
+            </div>
+            <div class="cancel">
+                <div class="btn btn-danger" (click)="backClicked()"> 
+                    go back
```
Comment (satyapriyamishra222) [inline src/app/components/external-link-access/external-link-access.component.html:15]: Translation.

File: src/app/components/external-link-access/external-link-access.component.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  selector: 'app-external-link-access',
+  templateUrl: './external-link-access.component.html',
+  styleUrls: ['./external-link-access.component.css']
+})
+export class ExternalLinkAccessComponent implements OnInit {
+
+  constructor(private activatedRoute: ActivatedRoute, private _location: Location) { }
+
+  public link: any;
+  ngOnInit(): void {
```
Comment (satyapriyamishra222) [inline src/app/components/external-link-access/external-link-access.component.ts:17]: void will not be necessary I guess ?

File: src/app/components/external-link-access/external-link-access.component.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+@Component({
+  selector: 'app-external-link-access',
+  templateUrl: './external-link-access.component.html',
+  styleUrls: ['./external-link-access.component.css']
+})
+export class ExternalLinkAccessComponent implements OnInit {
+
+  constructor(private activatedRoute: ActivatedRoute, private _location: Location) { }
+
+  public link: any;
```
Comment (satyapriyamishra222) [inline src/app/components/external-link-access/external-link-access.component.ts:16]: Can we provide the appropriate type here if it is not going to be type casted during the life cycle

File: src/app/components/mayastor-dialog/mayastor-dialog.component.ts [hunk 14]
```diff
@@ -9,21 +9,25 @@ import { ActivatedRoute, Router } from '@angular/router';
   styleUrls: ['./mayastor-dialog.component.scss']
 })
 export class MayastorDialogComponent implements OnInit {
-  public issueID: Number;
+  public issueID: any;
```
Comment (satyapriyamishra222) [inline src/app/components/mayastor-dialog/mayastor-dialog.component.ts:12]: May be a string data type or whichever is appropriate.

File: src/app/components/mayastor-dialog/mayastor-dialog.component.ts [hunk 15]
```diff
@@ -9,21 +9,25 @@ import { ActivatedRoute, Router } from '@angular/router';
...
   public testData: any;
+  passed: any;
```
Comment (satyapriyamishra222) [inline src/app/components/mayastor-dialog/mayastor-dialog.component.ts:14]: data type

Comment (satyapriyamishra222) [reply src/app/components/mayastor-dialog/mayastor-dialog.component.ts:14]: passed: number
failed: number

File: src/app/components/mayastor/mayastor.component.ts [hunk 16]
```diff
@@ -11,15 +13,28 @@ import { Status } from 'src/app/model/enum.model';
 })
 export class MayastorComponent implements OnInit {
   data: any;
-  constructor(private ApiService: DashboardData, private router: Router) {
-    ApiService.getMayastorTest().subscribe((res) => {
+  mySubscription: any;
+
+  constructor(private ApiService: DashboardData, private router: Router, private activatedRoute: ActivatedRoute) {
+    this.router.routeReuseStrategy.shouldReuseRoute = function () {
+      return false;
```
Comment (satyapriyamishra222) [inline src/app/components/mayastor/mayastor.component.ts:20]: What is the use/intention  here ?

File: src/app/components/mayastor/mayastor.component.ts [hunk 17]
```diff
@@ -32,5 +47,28 @@ export class MayastorComponent implements OnInit {
...
+  }
+  ngOnDestroy() {
+    if (this.mySubscription) {
+      this.mySubscription.unsubscribe();
+    }
+  }
+
+  goToURL(title, reference) {
+    let url: string;
+    if (title === 'Jira') {
```
Comment (satyapriyamishra222) [inline src/app/components/mayastor/mayastor.component.ts:63]: raw Strings here

File: src/assets/languages/default-en.json [hunk 18]
```diff
@@ -54,8 +54,8 @@
       "IMAGETAG":"ImageTag",
       "LENGTH":"Test count",
       "STATUS":"Status",
-      "STARTED_ON":"Statred At",
-      "JENKINS_URL":"Jenkins_link"
+      "STARTED_ON":"statred at",
```
Comment (satyapriyamishra222) [inline src/assets/languages/default-en.json:57]: spelling mistake here

File: src/app/components/doughnut-graph/doughnut-graph.component.ts (hunk 7 above)
Comment (bhaskarhc) [reply src/app/components/doughnut-graph/doughnut-graph.component.ts:127]: its doughnut radius fill metrics . if data not present then doughnut will be on red color

File: src/app/components/external-link-access/external-link-access.component.ts (hunk 12 above)
Comment (bhaskarhc) [reply src/app/components/external-link-access/external-link-access.component.ts:17]: yes, i think on latestAngular version or ESlint autofilling type of ng functions

File: src/app/components/external-link-access/external-link-access.component.ts (hunk 13 above)
Comment (bhaskarhc) [reply src/app/components/external-link-access/external-link-access.component.ts:16]: updated type as String

File: src/app/components/mayastor/mayastor.component.ts (hunk 16 above)
Comment (bhaskarhc) [reply src/app/components/mayastor/mayastor.component.ts:20]: /mayastor/{ID} this one will update the data on each route change .
If this one wont present , route will change and component data not updating .

File: src/assets/languages/default-en.json (hunk 18 above)
Comment (bhaskarhc) [reply src/assets/languages/default-en.json]: done

File: src/app/components/doughnut-graph/doughnut-graph.component.ts (hunk 3 above)
Comment (bhaskarhc) [reply src/app/components/doughnut-graph/doughnut-graph.component.ts:28]: rewriting this component . will change this complete structure

**Review by satyapriyamishra222 — APPROVED** [review]

**Pull request merged**